### PR TITLE
[ci] [python-package] check distributions with pydistcheck

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #5061.

Proposes adding `pydistcheck` to the set of checks run on Python sdists and wheels (the artifacts this project publishes to PyPI).

`pydistcheck` is a linter for Python distributions, similar with `check-wheel-contents` but with a very different set of checks, including:

* inclusion of debug symbols in `lib_lightgbm.so` / `lib_lightgbm.dll`
* accidental inclusion of unnecessary files
   - see #3639 and #3685, for example
* other portability issues documented at https://pydistcheck.readthedocs.io/en/latest/check-reference.html

I believe adding this will improve our confidence in releases of the Python package.

### Notes for Reviewers

Full disclosure, `pydistcheck` is my own personal side-project: https://github.com/jameslamb/pydistcheck. I started it after a conference talk I gave last year. I've resisted introducing it into LightGBM until now because I didn't think it was appropriate to use my position as a maintainer here to promote my own project.

I'm proposing it now because I think that project is at a level of maturity that it could genuinely be helpful here in LightGBM. I used it, for example, to catch two small packaging issues in `pandas` recently: 

* https://github.com/pandas-dev/pandas/issues/51900
* https://github.com/pandas-dev/pandas/pull/51931
* https://github.com/pandas-dev/pandas/pull/51971

And because it's been very helpful during my active development on changing the packaging strategy here for #5061 (e.g. in #5837 and #5759 ). It has helped me catch several mistakes during that development already.

Thanks very much for considering it.

### CODE REVIEW

File: .ci/check_python_dists.sh [hunk 1]
```diff
@@ -17,4 +18,12 @@ if { test "${TASK}" = "bdist" || test "${METHOD}" = "wheel"; }; then
     check-wheel-contents ${DIST_DIR}/*.whl || exit -1
 fi
 
+echo "pydistcheck..."
+pydistcheck \
+    --inspect \
```
Comment (jameslamb) [inline .ci/check_python_dists.sh:23]: This argument `--inspect` prints a summary like the following:

```text
----- package inspection summary -----
file size
  * compressed size: 1.3M
  * uncompressed size: 4.3M
  * compression space saving: 69.5%
contents
  * directories: 0
  * files: 16 (1 compiled)
size by extension
  * .so - 3963.7K (90.5%)
  * .py - 396.7K (9.1%)
  * no-extension - 17.6K (0.4%)
  * .txt - 0.0K (0.0%)
  * .typed - 0.0K (0.0%)
------------ check results -----------
errors found while checking: 0

```

([macOS bdist build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=14559&view=logs&j=52ea7a35-6bb0-5680-4def-de74fd7388e2&t=3a5da3c1-f658-52cd-3d24-edc674c4d20e))

Which I think would have been useful in situations like this: https://github.com/microsoft/LightGBM/pull/5252#pullrequestreview-1113112842

**Comment by hcho3** [comment]
@jameslamb This is interesting! Let me try pydistcheck on XGBoost. Right now, I often inspect the content of Python wheels by opening it as a ZIP archive. An automated check will be quite useful for me.

File: .ci/check_python_dists.sh [hunk 2]
```diff
@@ -17,4 +17,35 @@ if { test "${TASK}" = "bdist" || test "${METHOD}" = "wheel"; }; then
...
+if [ $PY_MINOR_VER -gt 7 ]; then
+    echo "pydistcheck..."
+    pip install pydistcheck
+    if [ $TASK == "CUDA" ] && [ $METHOD == "wheel" ]; then
+        pydistcheck \
+            --inspect \
+            --ignore 'compiled-objects-have-debug-symbols,max-allowed-size-compressed' \
+            --max-allowed-size-uncompressed '60M' \
+            --max-allowed-files 800 \
+            ${DIST_DIR}/* || exit -1
```
Comment (jameslamb) [inline .ci/check_python_dists.sh:30]: This `if` branch accounts for the fact that the CUDA 11.x wheel currently has the following warnings from `pydistcheck`:

```text
1. [compiled-objects-have-debug-symbols] Found compiled object containing debug symbols. For details, extract the distribution contents and run 'nm --debug-syms "lightgbm/lib_lightgbm.so"'.
2. [distro-too-large-compressed] Compressed size 35.1M is larger than the allowed size (5.0M).
3. [distro-too-large-uncompressed] Uncompressed size 54.8M is larger than the allowed size (15.0M).
errors found while checking: 3
```

([build link](https://github.com/microsoft/LightGBM/actions/runs/4739461720/jobs/8414327804?pr=5838))

I'm not sure why the CUDA build is so much larger than the non-CUDA one or where those debug symbols are coming from...maybe we are statically linking to CUDA libraries? I'm not sure.

Either way, I'm proposing not spending time investigating that right now, and instead just at least adding this check to detect if changes make the wheel significantly larger (which could be a problem in storage-sensitive environments like AWS Lambda).

File: .ci/check_python_dists.sh [hunk 3]
```diff
@@ -17,4 +17,35 @@ if { test "${TASK}" = "bdist" || test "${METHOD}" = "wheel"; }; then
...
+    elif [ $ARCH == "aarch64" ]; then
+        pydistcheck \
+            --inspect \
+            --ignore 'compiled-objects-have-debug-symbols' \
+            --max-allowed-size-compressed '5M' \
+            --max-allowed-size-uncompressed '15M' \
+            --max-allowed-files 800 \
+            ${DIST_DIR}/* || exit -1
```
Comment (jameslamb) [inline .ci/check_python_dists.sh:38]: The aarch64 wheel we produces one error from `pydistcheck`:

```text
1. [compiled-objects-have-debug-symbols] Found compiled object containing debug symbols. For details, extract the distribution contents and run 'nm --debug-syms "lightgbm/lib_lightgbm.so"'.
errors found while checking: 1
```

([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=14565&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9))

I'm not sure where those are coming from, but I'm proposing ignoring them for now. The aarch64 linux integrated OpenCL wheel has a very similar compressed size (2.3MB) to the arm64 linux wheels (1.6M).

**Review by jmoralez — APPROVED** [review]

**Pull request merged**